cg.go

```diff
@@ -0,0 +1,293 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartThreshold  = 0.9
+)
+
+// CGVariant represents the formula for computing the scaling parameter (in
+// literature usually denoted by the Greek letter β) used for updating the
+// conjugate direction in the nonlinear conjugate gradient (CG) method.
+type CGVariant interface {
+	// Init is called at the first iteration and provides a way to initialize
+	// any internal state.
+	Init(loc Location)
+	// Beta returns the value of the scaling parameter that is computed
+	// according to the particular variant of the CG method.
+	Beta(grad, gradPrev, dirPrev []float64) float64
+}
+
+// CG implements the nonlinear conjugate gradient method for solving nonlinear
+// unconstrained optimization problems [1]. It is a line search method that
+// generates the search directions d_k according to the formula
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
+// Different variants of the conjugate gradient method differ in the choice
+// of the parameter β_k. Conjugate gradient method requires no matrix storage
+// and is faster than the gradient descent method, but L-BFGS is usually more
```

Comment:
"Faster" and "more efficient" are misleading, because update flops is one metric,  and function evaluations are a totally different one.  Instead,  just say that CG usually requires fewer function evaluations,  and leave off words that sound like speed. I'm not sure what you mean by lbfgs being more robust, I've heard CG is more robust when the gradients are noisy